plugins/reflection/src/main/java/org/qbicc/plugin/reflection/ReflectionAnalyzingBasicBlockBuilder.java

```diff
@@ -0,0 +1,196 @@
+import org.qbicc.type.definition.LoadedTypeDefinition;
+import org.qbicc.type.definition.element.FieldElement;
+import org.qbicc.type.definition.element.MethodElement;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This BBB provides automatic registration/folding of reflective operations
+ * with constant arguments. Like GraalVM's "Automatic Detection"
+ * of reflection (https://www.graalvm.org/22.0/reference-manual/native-image/Reflection/),
+ * the goal is to support simple patterns for reflection without explicit user annotation.
+ */
+public class ReflectionAnalyzingBasicBlockBuilder extends DelegatingBasicBlockBuilder {
+    private static final AttachmentKey<KnownMethods> KEY = new AttachmentKey<>();
+
+
+    private final CompilationContext ctxt;
+
+    public ReflectionAnalyzingBasicBlockBuilder(final FactoryContext ctxt, final BasicBlockBuilder delegate) {
+        super(delegate);
+        this.ctxt = getContext();
+    }
+
+    @Override
+    public Value call(Value targetPtr, Value receiver, List<Value> arguments) {
+        Value replacement = process(targetPtr, receiver, arguments);
+        return replacement != null ? replacement : super.call(targetPtr, receiver, arguments);
+    }
+
+    @Override
+    public Value callNoSideEffects(Value targetPtr, Value receiver, List<Value> arguments) {
+        Value replacement = process(targetPtr, receiver, arguments);
+        return replacement != null ? replacement : super.callNoSideEffects(targetPtr, receiver, arguments);
+    }
+
+
+    @Override
+    public Value invoke(Value targetPtr, Value receiver, List<Value> arguments, BlockLabel catchLabel, BlockLabel resumeLabel, Map<Slot, Value> targetArguments) {
+        Value replacement = process(targetPtr, receiver, arguments);
+        return replacement != null ?  replacement : super.invoke(targetPtr, receiver, arguments, catchLabel, resumeLabel, targetArguments);
+    }
+
+    private Value process(Value targetPtr, Value receiver, List<Value> arguments) {
+        if (targetPtr instanceof ExecutableLiteral el) {
+            KnownMethods km = KnownMethods.get(ctxt);
+
+            // If the target is a method of java.lang.Class, then see if it is one of the reflection entry points
+            // we want to handle specially.
+            if (el.getExecutable() instanceof MethodElement me && me.getEnclosingType().load().equals(km.jlc)) {
+                // replace Class.forName(<literal String>) with a class literal if the type can be resolved at build time
+                if (me.equals(km.forName) && arguments.get(0) instanceof StringLiteral sl) {
+                    String internalName = sl.getValue().replace('.', '/');
+                    DefinedTypeDefinition dtd = getCurrentClassContext().findDefinedType(internalName);
+                    if (dtd != null) {
+                        // TODO: It would be really nice to simply replace the call to forName with a classOf(...).
```

Comment:
I feel like I'm missing something hopefully simple here.  Without doing the "constant fold" here, we don't actually get much benefit in terms of either optimization or analysis.  Posting the PR to get some ideas...

Reply:
Worked out a plan on Zulip: https://qbicc.zulipchat.com/#narrow/stream/284393-qbicc-development/topic/Class.20init/near/323749027